### PR TITLE
DRILL-4504: Create an event loop for each of [user, control, data] RPC components

### PR DESCRIPTION
**Pull request by sudheeshkatkam**

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/BootStrapContext.java [hunk 1]
```diff
@@ -50,8 +52,12 @@
   public BootStrapContext(DrillConfig config, ScanResult classpathScan) {
     this.config = config;
     this.classpathScan = classpathScan;
-    this.loop = TransportCheck.createEventLoopGroup(config.getInt(ExecConstants.BIT_SERVER_RPC_THREADS), "BitServer-");
-    this.loop2 = TransportCheck.createEventLoopGroup(config.getInt(ExecConstants.BIT_SERVER_RPC_THREADS), "BitClient-");
+    this.controlLoopGroup = TransportCheck.createEventLoopGroup(config.getInt(ExecConstants.BIT_SERVER_RPC_THREADS),
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/BootStrapContext.java]: We should use a different variable (with a much smaller number of threads)

File: exec/java-exec/src/test/java/org/apache/drill/exec/client/DrillClientSystemTest.java [hunk 2]
```diff
@@ -51,7 +51,7 @@ public void tearDownTest() {
   @Test
   public void testSubmitPlanSingleNode() throws Exception {
     startCluster(1);
-    DrillClient client = new DrillClient();
+    DrillClient client = DrillClient.newBuilder().build();
```
Comment (jacques-n) [inline exec/java-exec/src/test/java/org/apache/drill/exec/client/DrillClientSystemTest.java]: Let's keep the default constructor to be the same as the default builder. No reason to force a builder when there are no properties.

File: exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java [hunk 3]
```diff
@@ -74,73 +74,148 @@
...
+  // if the following variables are set during construction, they are not overridden during or after #connect call
+  // otherwise, they are set to defaults during #connect call
   private EventLoopGroup eventLoopGroup;
   private ExecutorService executor;
+  private boolean supportComplexTypes;
+
+  // the following variables are set during connection, and must not be overridden later
+  private UserClient client;
+  private UserProperties props;
+  private volatile ClusterCoordinator clusterCoordinator;
```
Comment (hnfgns) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: -0. 

Why volatile here? Is DrillClient meant to be thread safe? If so, we seem to have more work to do: #close for instance. Otherwise volatile seems totally irrelevant.

File: exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java [hunk 4]
```diff
@@ -74,73 +74,148 @@
...
+    this.ownsAllocator = ownsAllocator;
+    this.allocator = checkNotNull(allocator);
+    this.ownsZkConnection = ownsZkConnection;
+    this.clusterCoordinator = clusterCoordinator;
+    this.isDirectConnection = isDirectConnection;
+    this.ownsEventLoopGroup = ownsEventLoopGroup;
+    this.eventLoopGroup = eventLoopGroup;
+    this.ownsExecutor = ownsExecutor;
+    this.executor = executor;
+    this.supportComplexTypes = supportComplexTypes;
```
Comment (hnfgns) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: Nice work here. Would not it be nice if the builder provided "meaningful" defaults so that the complex creational patterns in this class particularly in #connect(...) could be eliminated further? For instance, why would executor ever be null or cluster coordinator be created later in the lifetime of this instance given instantiation cost is very minimal?

File: exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java (hunk 3 above)
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: I'll remove modifier, and document that the class is not thread safe.

Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: The query interface is expected to be threadsafe. Connect and close are not.

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/BootStrapContext.java (hunk 1 above)
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/BootStrapContext.java]: Done. I am hesitant to lower the control loop size until all blocking calls in handling requests are removed (e.g. cancelling an executor while fragment is starting up).

File: exec/java-exec/src/test/java/org/apache/drill/exec/client/DrillClientSystemTest.java (hunk 2 above)
Comment (sudheeshkatkam) [inline exec/java-exec/src/test/java/org/apache/drill/exec/client/DrillClientSystemTest.java]: Done.

File: exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java (hunk 3 above)
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: Updated the doc.

File: exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java (hunk 4 above)
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: I moved executor and eventLoopGroup instantiation to the c'tor. Initializing clusterCoordinator is tied to \#connect method, so I added a TODO as it requires more work.

**Comment by sudheeshkatkam** [comment]
Please re-review as I have more changes; I hope splitting up the changes into 4 commits helps.

@jacques-n can you also review rpc.control package info?

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/BootStrapContext.java (hunk 1 above)
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/BootStrapContext.java]: You're effectively doubling the number of threads with your current change. :/

**Comment by jacques-n** [comment]
The package info looks good and is a nice addition. Thanks!

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/BootStrapContext.java (hunk 1 above)
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/BootStrapContext.java]: Yes, I missed the obvious :) I'll change them to 5 each.

Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/BootStrapContext.java]: I'm a little nervous about decreasing the data pool that much. It might be wise to do some testing before decreasing it to 5.

**Comment by sudheeshkatkam** [comment]
There are two independent changes in this patch. So I moved the DrillClient changes to #480

**Comment by jacques-n** [comment]
Abandoned.
